File: Difficulty: Medium/Minimize the Heights I/minimize-the-heights-i.py
#User function Template for python3

class Solution:
    def getMinDiff(self,k, arr):
        # code here
        n = len(arr)
        if n == 1:
            return 0   

        
        arr.sort()

        
        ans = arr[-1] - arr[0]

         
        for i in range(1, n):
            
            max_height = max(arr[-1] - k, arr[i-1] + k)
            min_height = min(arr[0] + k, arr[i] - k)
            
            
            ans = min(ans, max_height - min_height)

        return ans



#{ 
 # Driver Code Starts
#Initial Template for Python 3
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        k = int(input())
        arr = list(map(int, input().strip().split()))
        solution = Solution()
        res = solution.getMinDiff(k, arr)
        print(res)
        print("~")

# } Driver Code Ends

File: Difficulty: Medium/Minimize the Heights I/minimize-the-heights-i.cpp
//{ Driver Code Starts
// Initial template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    int getMinDiff(int k, vector<int> &arr) {
        // code here
       
        sort(arr.begin(), arr.end());

        int n = arr.size();
        int ans = arr[n-1] - arr[0];
        int tempmin, tempmax;

        for (int i = 1; i < n; i++) {
        tempmin = min(arr[0] + k, arr[i]-k);
        tempmax = max(arr[n-1] - k, arr[i-1] + k);
        ans = min(ans, tempmax - tempmin);
        }

        return ans;
        }
    
    };

//{ Driver Code Starts.
int main() {
    int t;

    cin >> t;
    cin.ignore();

    while (t--) {
        int k;
        cin >> k;
        cin.ignore();
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        int n = arr.size();
        Solution ob;
        int res = ob.getMinDiff(k, arr);
        cout << res;

        cout << "\n";
        cout << "~"
             << "\n";
    }
    return 0;
}
// } Driver Code Ends